### PR TITLE
xdsclient: add a convenience type to synchronize execution of callbacks

### PR DESCRIPTION
**Pull request by easwars**
We need to guarantee that callbacks scheduled by the xDS client are executed in the order in which they were scheduled. Currently, this is not exactly the case and is the reason for https://github.com/grpc/grpc-go/issues/5429.

This type will get used by the xDS client as part of the refactor effort to support a resource agnostic watch API.

#resource-agnostic-xdsclient-api

RELEASE NOTES: none

### CODE REVIEW

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [inline xds/internal/xdsclient/callback_serializer.go:35]: Why is this exported?  Will other packages be using it?  If so, move it to somewhere else?  Maybe even `grpcsync` itself?

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [inline xds/internal/xdsclient/callback_serializer.go:60]: This is a race.

But does it matter?

Why not just skip the `if HasFired`?

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [inline xds/internal/xdsclient/callback_serializer.go:69]: This looks like a race to me.  Maybe just use a mutex and a bool?

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [inline xds/internal/xdsclient/callback_serializer.go:75]: `defer t.done.Fire()`?

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [inline xds/internal/xdsclient/callback_serializer.go:61]: It seems like nothing ever clears out `t.callbacks`.  Maybe this is okay if `t` is dereferenced, it should get garbage collected automatically?

Should `buffer.Unbounded` just have its own `Close` method that we leverage instead of needing another `closed` event here?

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:35]: I initially had a comment with a TODO saying move this to somewhere outside of the xds tree so that other packages will have access to it. But I can't think of anyone who needs it now. So, I will unexport it now, and we can always move it to some place appropriate when the time comes. Thanks.

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:75]: Done.

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:60]: Moved to a mutex with a bool.

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:61]: I think it would be nice to have a `Close()` or a `Drain()` method on `buffer.Unbounded`. After the recent memory leak that we have, I'm leaning towards being defensive and cleaning up as much as we can.

I could introduce such a function, and change the behavior of existing functions as follows:
- Put(): after close, would be a nop
- Load(): after close, would be a nop
- Get(): after close, would return a channel on which it is guaranteed that nothing would be sent

But there are a lot of users of `buffer.Unbounded` currently, and I can do this cleanup in a follow up PR, if you agree to the semantics change on the other methods as listed above.

Let me know what you think.

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:69]: Moved. Thanks.

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [reply xds/internal/xdsclient/callback_serializer.go:73]: Something like that seems plausible.  Let's discuss more offline.

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [inline xds/internal/xdsclient/callback_serializer.go:83]: Would it be a problem to just do `t.callbacks.Put(f)` and nothing else here?  What would be the harm in adding a callback if `run` had exited?  It would never be executed, and anyone still holding a ref to the `callbackSerializer` would be to blame for any leaked memory.  This would be true even without that behavior, since `Close` doesn't drain the buffer.  Theoretically if you call `Close` but don't dereference the struct right afterwards, you are doing it wrong.

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [inline xds/internal/xdsclient/callback_serializer.go:101]: It might be interesting to pass `ctx` to the callback; WDYT?

(If we did, maybe we don't even need to check `t.closed` at all here?  But that wouldn't be the main motivation.)

File: xds/internal/xdsclient/callback_serializer.go
Comment (dfawley) [inline xds/internal/xdsclient/callback_serializer.go:49]: What if this accepted a context as a parameter, and we let the user cancel the context to `Close` it?  Is that better or worse?  I don't know precisely the context in which this will be used, but if there's already a context this could simplify things.  If not, it could make them a little more complex (maybe?).

Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:49]: The callback serializer is currently used by `clientImpl` (to schedule watch callbacks for error conditions that happen even before it can determine the authority for the resource), and it is used by the `authority` type for scheduling watch callbacks for all cases (good response, bad response, resource is deleted etc).

Both these types don't have a context at this point in time. We _could_ add a context to these types, but that would probably be used only to be canceled at `Close()`, so that the callback serializer can cleanup. Creating a context, passing it to one type, and then canceling the context when we want to close that type seems cumbersome when compared to calling a `Close()` method.

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:101]: Passing a context to the callback does not guarantee that the callback respects cancelation of that context.

Also, this would mean a much bigger change to the xDS client API since we would have to change all watch callbacks (and their call sites) to accept a context. Most of the callbacks don't do blocking operations any way. So, passing a context to them would only make their life harder I think.

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:83]: Done.

**Comment by easwars** [comment]
@dfawley : I made the changes that we discussed around accepting a context when creating the serializer, and passing the same context around to callbacks, and relying on the context cancelation for closing the serializer. PTAL.

**Review by dfawley — COMMENTED** [review]
I like this form a lot better, as long as it doesn't make the lives of the callbacks that use it sufficiently harder than they needed to be.

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:40]: Done.

File: xds/internal/xdsclient/callback_serializer.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer.go:59]: Valid point. Done.

File: xds/internal/xdsclient/callback_serializer_test.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer_test.go:102]: The objective of this test is to ensure that all scheduled callbacks are executed. The objective of the previous test is to ensure that the callbacks are executed in the order in which they were scheduled. Technically, what is being tested here is also tested in the previous test. But the previous test is more complicated than this one because it needs to determine the scheduling order. So, if there is something broken only in the logic which ensures that all scheduled callbacks are executed, this test will be easier to use for debugging.

File: xds/internal/xdsclient/callback_serializer_test.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer_test.go:169]: Are you talking about the infinite for loop comment? That is handled now.

I had tried running this test 100k times locally and it did pass earlier. But that is no guarantee that it wouldn't flake in GA.

File: xds/internal/xdsclient/callback_serializer_test.go
Comment (easwars) [reply xds/internal/xdsclient/callback_serializer_test.go:218]: Yes, i think it was a little flaky early on, and I had to do all this dance. Simplified the test a little now.

File: xds/internal/xdsclient/callback_serializer_test.go
Comment (dfawley) [reply xds/internal/xdsclient/callback_serializer_test.go:101]: It seems like the only difference is the previous test does not sort the expected/actual callbacks executed.  If you want to keep it to make debugging more obvious, maybe refactor into a helper so the rest of the test isn't duplicated?

Comment (easwars) [reply xds/internal/xdsclient/callback_serializer_test.go:101]: The first test calls `Schedule()` while holding a lock. So, in the first test, calls to `Schedule()` happen one after the other. We need to guarantee this because that is the only way we can ensure that callbacks are executed in the order in which they are scheduled.

The second test calls `Schedule()` without holding a lock. So, this is where calls are `Schedule()` are happening in a truly concurrent fashion. And I feel it is important that we retain that test.

As far as refactoring them to weed out repeated code is concerned, the only thing that can be factored out into a common function is the logic where a couple of goroutines are spawned to capture the order of scheduling and execution. I tried making a function for this, but it was very ugly because of the number of arguments that need to be passed and the result parameters. 

Let me know if you think otherwise. Thanks.

Comment (dfawley) [reply xds/internal/xdsclient/callback_serializer_test.go:101]: Maybe simplify this test a bit further?  Why capture the execution order if you don't verify it?  How about just a waitgroup and make each scheduled callback execute `wg.Done()`?  If they are all executed, then `wg.Wait()` will eventually unblock and the test can pass.

Also maybe rename this to `TestCallbackSerializer_Schedule_ConcurrentCalls()` or something?

Comment (easwars) [reply xds/internal/xdsclient/callback_serializer_test.go]: Done. Thanks.

**Review by dfawley — APPROVED** [review]

**Pull request merged**
